src/mlpack/tests/dcgan_test.cpp

```diff
@@ -0,0 +1,288 @@
+
+  // Create the Discriminator network
+  FFN<SigmoidCrossEntropyError<> > discriminator;
+  discriminator.Add<Convolution<> >(1, dNumKernels, 4, 4, 2, 2, 1, 1, 28, 28);
+  discriminator.Add<LeakyReLU<> >(0.2);
+  discriminator.Add<Convolution<> >(dNumKernels, 2 * dNumKernels, 4, 4, 2, 2,
+      1, 1, 14, 14);
+  discriminator.Add<LeakyReLU<> >(0.2);
+  discriminator.Add<Convolution<> >(2 * dNumKernels, 4 * dNumKernels, 4, 4,
+      2, 2, 1, 1, 7, 7);
+  discriminator.Add<LeakyReLU<> >(0.2);
+  discriminator.Add<Convolution<> >(4 * dNumKernels, 8 * dNumKernels, 4, 4,
+      2, 2, 2, 2, 3, 3);
+  discriminator.Add<LeakyReLU<> >(0.2);
+  discriminator.Add<Convolution<> >(8 * dNumKernels, 1, 4, 4, 1, 1,
+      1, 1, 2, 2);
+  discriminator.Add<SigmoidLayer<> >();
+
+  // Create the Generator network
+  FFN<SigmoidCrossEntropyError<> > generator;
+  generator.Add<TransposedConvolution<> >(noiseDim, 8 * dNumKernels, 2, 2,
+      1, 1, 1, 1, 1, 1);
+  generator.Add<ReLULayer<> >();
+  generator.Add<TransposedConvolution<> >(8 * dNumKernels, 4 * dNumKernels,
+      2, 2, 1, 1, 0, 0, 2, 2);
+  generator.Add<ReLULayer<> >();
+  generator.Add<TransposedConvolution<> >(4 * dNumKernels, 2 * dNumKernels,
+      5, 5, 2, 2, 1, 1, 3, 3);
+  generator.Add<ReLULayer<> >();
+  generator.Add<TransposedConvolution<> >(2 * dNumKernels, dNumKernels, 8, 8,
+      1, 1, 1, 1, 7, 7);
+  generator.Add<ReLULayer<> >();
+  generator.Add<TransposedConvolution<> >(dNumKernels, 1, 15, 15, 1, 1, 1, 1,
+      14, 14);
+  generator.Add<TanHLayer<> >();
+
+  // Create GAN
+  GaussianInitialization gaussian(0, 1);
+  Adam optimizer(stepSize, batchSize, 0.9, 0.999, eps, numIterations,
+      tolerance, shuffle);
+  std::function<double()> noiseFunction = [] () {
+      return math::RandNormal(0, 1);};
+  GAN<FFN<SigmoidCrossEntropyError<> >, GaussianInitialization,
+      std::function<double()> > gan(trainData, generator, discriminator,
+      gaussian, noiseFunction, noiseDim, batchSize, generatorUpdateStep,
+      discriminatorPreTrain, multiplier);
+
+  std::cout << "Training..." << std::endl;
+  gan.Train(optimizer);
+
+  // Generate samples
+  std::cout << "Sampling..." << std::endl;
+  arma::mat noise(noiseDim, 1);
+  size_t dim = std::sqrt(trainData.n_rows);
+  arma::mat generatedData(2 * dim, dim * numSamples);
+
+  for (size_t i = 0; i < numSamples; i++)
+  {
+    arma::mat samples;
+    noise.imbue( [&]() { return noiseFunction(); } );
+
+    generator.Forward(noise, samples);
+    samples.reshape(dim, dim);
+    samples = samples.t();
+
+    generatedData.submat(0, i * dim, dim - 1, i * dim + dim - 1) = samples;
+
+    samples = trainData.col(math::RandInt(0, trainData.n_cols));
+    samples.reshape(dim, dim);
+    samples = samples.t();
+
+    generatedData.submat(dim,
+        i * dim, 2 * dim - 1, i * dim + dim - 1) = samples;
+  }
+
+  std::cout << "Saving output to " << output_dataset << "..." << std::endl;
+  generatedData.save(output_dataset, arma::csv_ascii);
```

Comment:
In an automated test case, I think we don't need to save the output, but we can comment the lines, and add a comment; I guess someone might find this helpful. Also, I'd like to note that we just test if the the code builds and runs, since we do not take a look at the output.

Reply:
If we just use this one to see if the code builds, perhaps we can decrease the number of training iterations?

Reply:
The number of iterations is anyways down to the minimum (`1`).